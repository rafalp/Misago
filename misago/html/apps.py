from django.apps import AppConfig


class MisagoHTMLConfig(AppConfig):
    name = "misago.html"
    label = "misago_HTML"
    verbose_name = "Misago HTML"
