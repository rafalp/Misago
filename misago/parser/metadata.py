from typing import Iterable

from django.conf import settings
from django.contrib.auth import get_user_model

from ..core.utils import slugify
from .context import ParserContext
from .hooks import (
    get_ast_metadata_users_queryset_hook,
    update_ast_metadata_hook,
    update_ast_metadata_from_node_hook,
    update_ast_metadata_users_hook,
)

User = get_user_model()


def create_ast_metadata(
    context: ParserContext,
    ast: list[dict],
) -> dict:
    metadata = {
        "outbound-links": set(),
        "usernames": set(),
        "users": {},
        "posts": {
            "ids": set(),
            "objs": {},
        },
    }

    return update_ast_metadata_hook(_update_ast_metadata_action, context, ast, metadata)


def _update_ast_metadata_action(
    context: ParserContext,
    ast: list[dict],
    metadata: dict,
) -> dict:
    for ast_node in ast:
        update_ast_metadata_from_node(context, ast_node, metadata)

    update_ast_metadata_users(context, metadata)

    return metadata


def update_ast_metadata_from_node(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
) -> None:
    update_ast_metadata_from_node_hook(
        _update_ast_metadata_from_node_action, context, ast_node, metadata
    )


def _update_ast_metadata_from_node_action(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
) -> None:
    if ast_node["type"] == "mention":
        metadata["usernames"].add(slugify(ast_node["username"]))

    elif ast_node["type"] == "quote-bbcode":
        if ast_node["author"]:
            metadata["usernames"].add(slugify(ast_node["author"]))
        if ast_node["post"]:
            metadata["posts"]["ids"].add(ast_node["post"])

    elif ast_node["type"] in ("auto-link", "auto-url", "url", "url-bbcode"):
        if not context.forum_address.is_inbound_link(ast_node["href"]):
            metadata["outbound-links"].add(ast_node["href"])

    if ast_node.get("children"):
        for child_node in ast_node["children"]:
            update_ast_metadata_from_node(context, child_node, metadata)

    if ast_node.get("items"):
        for child_node in ast_node["items"]:
            update_ast_metadata_from_node(context, child_node, metadata)

    if ast_node.get("lists"):
        for child_node in ast_node["lists"]:
            update_ast_metadata_from_node(context, child_node, metadata)


def update_ast_metadata_users(context: ParserContext, metadata: dict) -> None:
    update_ast_metadata_users_hook(_update_ast_metadata_users_action, context, metadata)


def _update_ast_metadata_users_action(context: ParserContext, metadata: dict) -> None:
    if not metadata["usernames"]:
        return

    usernames = sorted(metadata["usernames"])
    if len(usernames) > settings.MISAGO_PARSER_MAX_USERS:
        return

    if usernames:
        queryset = get_ast_metadata_users_queryset(context, usernames)
        for user in queryset:
            metadata["users"][user.slug] = user


def get_ast_metadata_users_queryset(
    context: ParserContext, usernames: list[str]
) -> Iterable[User]:
    return get_ast_metadata_users_queryset_hook(
        _get_ast_metadata_users_queryset_action, context, usernames
    )


def _get_ast_metadata_users_queryset_action(
    context: ParserContext, usernames: list[str]
) -> Iterable[User]:
    return User.objects.filter(slug__in=usernames)
