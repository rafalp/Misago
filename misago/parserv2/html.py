from markdown_it import MarkdownIt
from markdown_it.token import Token


def render_tokens_to_html(parser: MarkdownIt, tokens: list[Token]) -> str:
    return parser.renderer.render(tokens, parser.options, {}).strip()
