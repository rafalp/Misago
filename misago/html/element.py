from html import escape


def html_element(
    tag: str,
    children: str | None = None,
    attrs: dict | None = None,
) -> str:
    html = tag

    if attrs:
        html += _html_attributes(attrs)

    if children is None:
        return f"<{html} />"

    return f"<{html}>{children}</{tag}>"


def _html_attributes(attrs: dict) -> str:
    html = ""
    for attr, value in attrs.items():
        if value is True:
            html += " " + attr
        elif value is not None:
            html += f' {attr}="{escape(value)}"'
    return html
