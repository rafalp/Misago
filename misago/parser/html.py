from html import escape

from django.utils.translation import pgettext

from ..core.utils import slugify
from .context import ParserContext
from .exceptions import AstError
from .hooks import complete_markup_html_hook, render_ast_node_to_html_hook
from .urls import clean_href

SPOILER_SUMMARY = "<spoiler-summary-message>"


def render_ast_to_html(context: ParserContext, ast: list[dict], metadata: dict) -> str:
    html = ""
    for ast_node in ast:
        node_html = render_ast_node_to_html(context, ast_node, metadata)
        if node_html:
            html += node_html
    return html


def render_ast_node_to_html(
    context: ParserContext, ast_node: dict, metadata: dict
) -> str:
    return render_ast_node_to_html_hook(
        _render_ast_node_to_html_action, context, ast_node, metadata
    )


def _render_ast_node_to_html_action(
    context: ParserContext, ast_node: dict, metadata: dict
) -> str:
    ast_type = ast_node["type"]

    if ast_type in ("heading", "heading-setex"):
        html_tag = f"h{ast_node['level']}"
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<{html_tag}>{children}</{html_tag}>"

    if ast_type == "list":
        html_tag = "ol" if ast_node["ordered"] else "ul"
        children = render_ast_to_html(context, ast_node["items"], metadata)
        return f"<{html_tag}>{children}</{html_tag}>"

    if ast_type == "list-item":
        text = render_ast_to_html(context, ast_node["children"], metadata)
        children = render_ast_to_html(context, ast_node["lists"], metadata)
        return f"<li>{text}{children}</li>"

    if ast_type in ("code", "code-bbcode"):
        if ast_node["syntax"]:
            html_class = f" class=\"language-{ast_node['syntax']}\""
        else:
            html_class = ""
        return f"<pre{html_class}><code>{escape(ast_node['code'])}</code></pre>"

    if ast_type == "code-indented":
        return f"<pre><code>{escape(ast_node['code'])}</code></pre>"

    if ast_type == "code-inline":
        return f"<code>{escape(ast_node['code'])}</code>"

    if ast_type == "quote":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<blockquote>{children}</blockquote>"

    if ast_type == "quote-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)

        if not ast_node["author"]:
            return f"<blockquote>{children}</blockquote>"

        heading = escape(ast_node["author"])

        return (
            '<aside class="quote-block">'
            f'<div class="quote-heading" data-noquote="1">{heading}</div>'
            f'<blockquote class="quote-body">{children}</blockquote>'
            "</aside>"
        )

    if ast_type == "spoiler-bbcode":
        if ast_node["summary"]:
            summary = escape(ast_node["summary"])
        else:
            summary = SPOILER_SUMMARY
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<details><summary>{summary}</summary>{children}</details>"

    if ast_type == "paragraph":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<p>{children}</p>"

    if ast_type in ("emphasis", "emphasis-underscore"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<em>{children}</em>"

    if ast_type in ("strong", "strong-underscore"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<strong>{children}</strong>"

    if ast_type in ("strikethrough", "strikethrough-bbcode"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<del>{children}</del>"

    if ast_type == "bold-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<b>{children}</b>"

    if ast_type == "italics-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<i>{children}</i>"

    if ast_type == "underline-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<u>{children}</u>"

    if ast_type == "superscript-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<sup>{children}</sup>"

    if ast_type == "subscript-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<sub>{children}</sub>"

    if ast_type in ("thematic-break", "thematic-break-bbcode"):
        return "<hr />"

    if ast_type in ("image", "image-bbcode"):
        src = escape(clean_href(ast_node["src"]))
        alt = escape(ast_node["alt"]) if ast_node["alt"] else ""
        return f'<img src="{src}" alt="{alt}" />'

    if ast_type in ("url", "url-bbcode"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        href = escape(clean_href(ast_node["href"]))
        rel = "external nofollow noopener"
        return f'<a href="{href}" rel="{rel}" target="_blank">{children or href}</a>'

    if ast_type in ("auto-link", "auto-url"):
        href = escape(clean_href(ast_node["href"]))
        rel = "external nofollow noopener"
        if ast_node.get("image"):
            return f'<img src="{href}" alt="" />'
        else:
            return f'<a href="{href}" rel="{rel}" target="_blank">{href}</a>'

    if ast_type == "mention":
        username = slugify(ast_node["username"])
        if username not in metadata["users"]:
            return escape("@" + ast_node["username"])

        user = metadata["users"][username]
        return f'<a href="{user.get_absolute_url()}">@{escape(user.username)}</a>'

    if ast_type == "escape":
        return escape(ast_node["character"])

    if ast_type == "line-break":
        return "<br />"

    if ast_type == "text":
        return escape(ast_node["text"])

    raise AstError(f"Unknown AST node type: {ast_type}")


def complete_markup_html(html: str, **kwargs) -> str:
    return complete_markup_html_hook(_complete_markup_html_action, html, **kwargs)


def _complete_markup_html_action(html: str, **kwargs) -> str:
    html = complete_markup_html_spoiler_summary(html)
    return html


def complete_markup_html_spoiler_summary(html: str) -> str:
    if SPOILER_SUMMARY in html:
        html = html.replace(
            SPOILER_SUMMARY, pgettext("spoiler summary", "Reveal spoiler")
        )

    return html
