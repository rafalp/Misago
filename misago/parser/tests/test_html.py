import pytest

from ..html import render_ast_to_html
from ..metadata import create_ast_metadata


@pytest.mark.parametrize(
    "markup",
    (
        "# Hello world!",
        "## Hello world!",
        "### Hello world!",
        "#### Hello world!",
        "##### Hello world!",
        "###### Hello world!",
    ),
)
def test_render_ast_to_html_heading(markup, parser_context, parse_markup, snapshot):
    ast = parse_markup(markup)
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


@pytest.mark.parametrize(
    "markup",
    (
        "Hello world!\n=====",
        "Hello world!\n-----",
    ),
)
def test_render_ast_to_html_setex_heading(
    markup, parser_context, parse_markup, snapshot
):
    ast = parse_markup(markup)
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_quote(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        > Hello world!
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_quote_bbcode(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        [quote]Hello world![/quote]
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_quote_bbcode_with_author(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        """
        [quote=Author]Hello world![/quote]
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_unordered_list(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        - Lorem
        - _Ipsum_
        - Dolor
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_ordered_list(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        1. Lorem
        2. _Ipsum_
        3. Dolor
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_unordered_list_with_nested_list(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        """
        - Lorem
        - _Ipsum_
          - Met
          - Elit
        - Dolor
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_minimal_table(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        | Col1 |
        | ---- |
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_table(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        | Col1 | Col2 |*Col3*| Col4 |
        | ---- | :--- | :--: | ---: |
        | Cel1 | Cel2 |      | Cel4 |
        | Cel5 |*Cel6*| Cel7 | Cel8 |
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_spoiler(parser_context, parse_markup, snapshot):
    ast = parse_markup("[spoiler]Hello world![/spoiler]")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_spoiler_with_summary(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("[spoiler=Secret message]Hello world![/spoiler]")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_paragraph(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello world!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_two_paragraphs(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello world!\n\nHow's going?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_thematic_break(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello world!\n- - -\nHow's going?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_thematic_break_bbcode(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello world!\n[hr]\nHow's going?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_code(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        ```
        if random.randint(0, 10) > 4:
            print("Gotcha!")
            return True
        return False
        ```
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_code_with_syntax(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        ```python
        if random.randint(0, 10) > 4:
            print("Gotcha!")
            return True
        return False
        ```
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_code_bbcode(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        [code]
            if random.randint(0, 10) > 4:
                print("Gotcha!")
                return True
            return False
        [/code]
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_code_bbcode_with_syntax(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        """
        [code=python]
            if random.randint(0, 10) > 4:
                print("Gotcha!")
                return True
            return False
        [/code]
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_code_indented(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        Hello:

            if random.randint(0, 10) > 4:
                print("Gotcha!")
                return True
            return False
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_inline_code(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello `world`!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_emphasis_text(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello *world*!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_emphasis_underscore_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello _world_!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_strong_text(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello **world**!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_strikethrough_text(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello ~~world~~!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_bold_bbcode_text(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello [b]world[/b]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_italics_bbcode_text(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello [i]world[/i]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_underline_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [u]world[/u]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_strikethrough_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [s]world[/s]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_superscript_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [sup]world[/sup]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_subscript_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [sub]world[/sub]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_escaped_characters(parser_context, parse_markup, snapshot):
    ast = parse_markup(r"Hello\[hr\]World!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_soft_linebreak(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello world!\nHow's going?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_image_auto_link(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the logo: <!https://misago-project.org/img.png>")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_image(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the logo: !(https://misago-project.org/img.png)")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_image_with_alt_text(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the logo: ![Alt text](https://misago-project.org/img.png)")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_image_with_title(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        f'See the logo: !(https://misago-project.org/img.png "Amazing, right?")'
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_image_with_title_and_alt_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        'See the logo: ![Alt text](https://misago-project.org/img.png "Amazing, right?")'
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_image_bbcode(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the logo: [img]https://misago-project.org/img.png[/img]")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_image_bbcode_with_alt_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        f"See the logo: [img=https://misago-project.org/img.png]Alt text[/img]"
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_url(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See [*the site*](https://misago-project.org)!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_plaintext_url_with_title(parser_context, parse_markup, snapshot):
    ast = parse_markup('See [*the site*](https://misago-project.org "misago forums")!')
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_url_bbcode(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See [url]https://misago-project.org[/url]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_url_bbcode_with_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See [url=https://misago-project.org]*the site*[/url]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_attachment(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the site: <attachment=image.png:123>")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_auto_link(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the file: <https://misago-project.org>")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_auto_url(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the site: https://misago-project.org")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_mention(parser_context, parse_markup, snapshot, user):
    user.id = 1000
    user.set_username("MentionedUser")
    user.set_email("mentioned@example.com")
    user.save()

    ast = parse_markup(f"How's going, @{user.username}?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_mention_non_existing_user(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"How's going, @JohnDoe?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_html(parser_context, ast, metadata)


def test_render_ast_to_html_for_unsupported_ast_raises_error(parser_context):
    with pytest.raises(ValueError):
        metadata = create_ast_metadata(parser_context, [])
        render_ast_to_html(parser_context, [{"type": "unsupported"}], metadata)
