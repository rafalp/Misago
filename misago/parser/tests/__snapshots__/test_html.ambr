# serializer version: 1
# name: test_complete_markup_html_replaces_default_spoiler_summary
  '<details><summary>Reveal spoiler</summary><p>Hello world!</p></details>'
# ---
# name: test_render_ast_to_html_auto_link
  '<p>See the site: <a href="https://misago-project.org" rel="external nofollow noopener" target="_blank">https://misago-project.org</a></p>'
# ---
# name: test_render_ast_to_html_auto_url
  '<p>See the site: <a href="https://misago-project.org" rel="external nofollow noopener" target="_blank">https://misago-project.org</a></p>'
# ---
# name: test_render_ast_to_html_bold_bbcode_text
  '<p>Hello <b>world</b>!</p>'
# ---
# name: test_render_ast_to_html_code
  '''
  <pre><code>if random.randint(0, 10) &gt; 4:
      print(&quot;Gotcha!&quot;)
      return True
  return False</code></pre>
  '''
# ---
# name: test_render_ast_to_html_code_bbcode
  '''
  <pre><code>if random.randint(0, 10) &gt; 4:
      print(&quot;Gotcha!&quot;)
      return True
  return False</code></pre>
  '''
# ---
# name: test_render_ast_to_html_code_bbcode_with_syntax
  '''
  <pre class="language-python"><code>if random.randint(0, 10) &gt; 4:
      print(&quot;Gotcha!&quot;)
      return True
  return False</code></pre>
  '''
# ---
# name: test_render_ast_to_html_code_indented
  '''
  <p>Hello:</p><pre><code>if random.randint(0, 10) &gt; 4:
      print(&quot;Gotcha!&quot;)
      return True
  return False</code></pre>
  '''
# ---
# name: test_render_ast_to_html_code_with_syntax
  '''
  <pre class="language-python"><code>if random.randint(0, 10) &gt; 4:
      print(&quot;Gotcha!&quot;)
      return True
  return False</code></pre>
  '''
# ---
# name: test_render_ast_to_html_emphasis_text
  '<p>Hello <em>world</em>!</p>'
# ---
# name: test_render_ast_to_html_emphasis_underscore_text
  '<p>Hello <em>world</em>!</p>'
# ---
# name: test_render_ast_to_html_escaped_characters
  '<p>Hello[hr]World!</p>'
# ---
# name: test_render_ast_to_html_heading[# Hello world!]
  '<h1>Hello world!</h1>'
# ---
# name: test_render_ast_to_html_heading[## Hello world!]
  '<h2>Hello world!</h2>'
# ---
# name: test_render_ast_to_html_heading[### Hello world!]
  '<h3>Hello world!</h3>'
# ---
# name: test_render_ast_to_html_heading[#### Hello world!]
  '<h4>Hello world!</h4>'
# ---
# name: test_render_ast_to_html_heading[##### Hello world!]
  '<h5>Hello world!</h5>'
# ---
# name: test_render_ast_to_html_heading[###### Hello world!]
  '<h6>Hello world!</h6>'
# ---
# name: test_render_ast_to_html_image
  '<p>See the logo: <img src="https://misago-project.org/img.png" alt="" /></p>'
# ---
# name: test_render_ast_to_html_image_auto_link
  '<p>See the logo: <img src="https://misago-project.org/img.png" alt="" /></p>'
# ---
# name: test_render_ast_to_html_image_bbcode
  '<p>See the logo: <img src="https://misago-project.org/img.png" alt="" /></p>'
# ---
# name: test_render_ast_to_html_image_bbcode_with_alt_text
  '<p>See the logo: <img src="https://misago-project.org/img.png" alt="Alt text" /></p>'
# ---
# name: test_render_ast_to_html_image_with_alt_text
  '<p>See the logo: <img src="https://misago-project.org/img.png" alt="Alt text" /></p>'
# ---
# name: test_render_ast_to_html_inline_code
  '<p>Hello <code>world</code>!</p>'
# ---
# name: test_render_ast_to_html_italics_bbcode_text
  '<p>Hello <i>world</i>!</p>'
# ---
# name: test_render_ast_to_html_mention
  '<p>How&#x27;s going, <a href="/u/mentioneduser/1000/">@MentionedUser</a>?</p>'
# ---
# name: test_render_ast_to_html_mention_non_existing_user
  '<p>How&#x27;s going, @JohnDoe?</p>'
# ---
# name: test_render_ast_to_html_ordered_list
  '<ol><li>Lorem</li><li><em>Ipsum</em></li><li>Dolor</li></ol>'
# ---
# name: test_render_ast_to_html_paragraph
  '<p>Hello world!</p>'
# ---
# name: test_render_ast_to_html_quote
  '<blockquote><p>Hello world!</p></blockquote>'
# ---
# name: test_render_ast_to_html_quote_bbcode
  '<blockquote><p>Hello world!</p></blockquote>'
# ---
# name: test_render_ast_to_html_quote_bbcode_with_author
  '<aside class="quote-block"><div class="quote-heading" data-noquote="1">Author</div><blockquote class="quote-body"><p>Hello world!</p></blockquote></aside>'
# ---
# name: test_render_ast_to_html_setex_heading[Hello world!\n-----]
  '<h2>Hello world!</h2>'
# ---
# name: test_render_ast_to_html_setex_heading[Hello world!\n=====]
  '<h1>Hello world!</h1>'
# ---
# name: test_render_ast_to_html_soft_linebreak
  '<p>Hello world!<br />How&#x27;s going?</p>'
# ---
# name: test_render_ast_to_html_spoiler
  '<details><summary><spoiler-summary-message></summary><p>Hello world!</p></details>'
# ---
# name: test_render_ast_to_html_spoiler_with_summary
  '<details><summary>Secret message</summary><p>Hello world!</p></details>'
# ---
# name: test_render_ast_to_html_strikethrough_bbcode_text
  '<p>Hello <del>world</del>!</p>'
# ---
# name: test_render_ast_to_html_strikethrough_text
  '<p>Hello <del>world</del>!</p>'
# ---
# name: test_render_ast_to_html_strong_text
  '<p>Hello <strong>world</strong>!</p>'
# ---
# name: test_render_ast_to_html_subscript_bbcode_text
  '<p>Hello <sub>world</sub>!</p>'
# ---
# name: test_render_ast_to_html_superscript_bbcode_text
  '<p>Hello <sup>world</sup>!</p>'
# ---
# name: test_render_ast_to_html_thematic_break
  '<p>Hello world!</p><hr /><p>How&#x27;s going?</p>'
# ---
# name: test_render_ast_to_html_thematic_break_bbcode
  '<p>Hello world!</p><hr /><p>How&#x27;s going?</p>'
# ---
# name: test_render_ast_to_html_two_paragraphs
  '<p>Hello world!</p><p>How&#x27;s going?</p>'
# ---
# name: test_render_ast_to_html_underline_bbcode_text
  '<p>Hello <u>world</u>!</p>'
# ---
# name: test_render_ast_to_html_unordered_list
  '<ul><li>Lorem</li><li><em>Ipsum</em></li><li>Dolor</li></ul>'
# ---
# name: test_render_ast_to_html_unordered_list_with_nested_list
  '<ul><li>Lorem</li><li><em>Ipsum</em><ul><li>Met</li><li>Elit</li></ul></li><li>Dolor</li></ul>'
# ---
# name: test_render_ast_to_html_url
  '<p>See <a href="https://misago-project.org" rel="external nofollow noopener" target="_blank"><em>the site</em></a>!</p>'
# ---
# name: test_render_ast_to_html_url_bbcode
  '<p>See <a href="https://misago-project.org" rel="external nofollow noopener" target="_blank">https://misago-project.org</a>!</p>'
# ---
# name: test_render_ast_to_html_url_bbcode_with_text
  '<p>See <a href="https://misago-project.org" rel="external nofollow noopener" target="_blank"><em>the site</em></a>!</p>'
# ---
